Source/Methods/RFmxDPD/RFmxDPD.cs

```diff
@@ -82,92 +81,115 @@ public static MemoryPolynomialConfiguration GetDefault()
             //Instantiate new waveform with reference waveform properties
-            mPResults.PostDpdWaveform = waveform;
-            RFmxSpecAnMXDpdApplyDpdIdleDurationPresent idlePresent = waveform.IdleDurationPresent ? RFmxSpecAnMXDpdApplyDpdIdleDurationPresent.True : RFmxSpecAnMXDpdApplyDpdIdleDurationPresent.False;
+            MemoryPolynomialResults mPResults = new MemoryPolynomialResults()
+            {
+                PostDpdWaveform = referenceWaveform
+            };
+            mPResults.PostDpdWaveform.WaveformName = referenceWaveform.WaveformName + "postMpDpd";
+            RFmxSpecAnMXDpdApplyDpdIdleDurationPresent idlePresent = referenceWaveform.IdleDurationPresent ? RFmxSpecAnMXDpdApplyDpdIdleDurationPresent.True : RFmxSpecAnMXDpdApplyDpdIdleDurationPresent.False;
+
+            RfsgGenerationStatus preDpdGenerationStatus = rfsgSession.CheckGenerationStatus();
+            rfsgSession.Abort(); // abort so we don't mess with the loop logic
+
             if (mpConfig.IterativeDpdEnabled == RFmxSpecAnMXDpdIterativeDpdEnabled.False)
-            {
                 mpConfig.NumberOfIterations = 1;
-            }
             for (int i = 0; i < mpConfig.NumberOfIterations; i++)
             {
-                specAnSignal.Dpd.Configuration.ConfigurePreviousDpdPolynomial(selectorString, mPResults.DpdPolynomial);
-                specAnSignal.Initiate("", selectorString);
+                specAn.Dpd.Configuration.ConfigurePreviousDpdPolynomial(selectorString, mPResults.DpdPolynomial);
+                rfsgSession.Initiate();
+                specAn.Initiate(selectorString, "");
                 //waveform data and PAPR are overwritten in post DPD waveform
-                specAnSignal.Dpd.ApplyDpd.ApplyDigitalPredistortion(selectorString, waveform.WaveformData, idlePresent, timeout_s, ref mPResults.PostDpdWaveform.WaveformData,
-                    out mPResults.PostDpdWaveform.PAPR_dB, out mPResults.PowerOffset_dB);
-                SG.DownloadWaveform(rfsgSession, mPResults.PostDpdWaveform);
+                specAn.Dpd.ApplyDpd.ApplyDigitalPredistortion(selectorString, referenceWaveform.WaveformData, idlePresent, 10.0, ref mPResults.PostDpdWaveform.WaveformData,
+                    out mPResults.PostDpdWaveform.PAPR_dB, out mPResults.PowerOffset_dB);
+                DownloadWaveform(rfsgSession, mPResults.PostDpdWaveform); // implicit abort
                 rfsgSession.RF.PowerLevel = rfsgSession.RF.PowerLevel + mPResults.PowerOffset_dB;
-                rfsgSession.Initiate();
-                specAnSignal.Dpd.Results.FetchDpdPolynomial(selectorString, timeout_s, ref mPResults.DpdPolynomial);
+                ConfigureWaveformToGenerate(rfsgSession, mPResults.PostDpdWaveform);
+                specAn.Dpd.Results.FetchDpdPolynomial(selectorString, 10.0, ref mPResults.DpdPolynomial);
             }
+
+            if (preDpdGenerationStatus == RfsgGenerationStatus.InProgress)
```

Comment:
How is this exactly working? Are re-writing the script if the rfsg was already running and then you restart?

Reply:
I just figured it might be nice to put the generator back into the state it was in before the function was called. For instance, if it was generating, then we make sure it is generating before we return. If it wasn't, well, vice-versa. 
Any criticism there?

Reply:
Resolving since this is now in a separate commit